hera_cal/red_groups.py

```diff
@@ -0,0 +1,246 @@
+                        
+        # TODO: there's almost certainly a quicker way to do this where we use all
+        #       the dict info that we have. But this will always work.
+        new_antpos = {**self.antpos, **other.antpos}
+        return RedundantGroups(
+            antpos = new_antpos,
+            bl_error_tol = self.bl_error_tol
+        )
+    
+    def with_additional_ants(self, antpos: dict[int, np.ndarray]):
+        # First check if we're actually adding anything.
+        if all(k in self.antpos for k in antpos):
+            return self
+        if self.antpos is None:
+            raise ValueError("can't add antennas to a RedundantGroups object without antpos")
+        
+        new_antpos = {**self.antpos, **antpos}
+        return RedundantGroups(
+            antpos = new_antpos,
+            bl_error_tol = self.bl_error_tol
+        )
+    
+    def append(self, red: Sequence[AntPair | Baseline]) -> None:
+        """In-place append a new redundant group to the list of redundant groups.
+        
+        This maintains the list-of-lists duck-typing of the redundant groups.
+        """
+        # First check if we're actually adding anything new
+        red = [r[:2] for r in red]
+
+        ubls = {r: self.get_ubl_key(r) for r in red if r in self}
+        
+        
+        if ubls:
+            
+            if len(set(ubls.values())) > 1:
+                invubl = {v: k for k, v in ubls.items()}
+                raise ValueError(
+                    "Attempting to add a new redundant group where some baselines "
+                    "already exist in differing groups. Got the following baselines "
+                    f"mapping to different groups: {dict(zip(invubl.values(), invubl.keys()))}"
+                )
+            if len(ubls) == len(red):
+                # already have all these!
+                # Note, we have to do this *after* the above check, because although
+                # we have all the baselines, they might not be in the same group, and 
+                # we want to raise an error in that case.
+                return
+
+            # Now we're gaurenteed that we have some baselines that are new, and that
+            # they all map to the same already-existing group.
+            ublkey = next(iter(ubls.values()))
+            self[ublkey] = sorted(set(self[ublkey] + red))
+            
+        else:
+            # We're adding a completely new group
+            self._red_list.append(red)
+            
+        # Reset the maps
+        del self._bl_to_red_map
+        del self._red_key_to_bls_map
+
+    def __len__(self):
+        return len(self._red_list)
+
+    def __iter__(self):
+        return iter(self._red_list)
+    
+    def __getitem__(self, key: int | AntPair | Baseline) -> list[AntPair | Baseline]:
+        if isinstance(key, int):
+            return self._red_list[key]
+        else:
+            return self.get_red(key)
+        
```

Comment:
I'd like to see this object support redcal.filter_reds performed on itself.

Reply:
Sure!